src/main/java/com/firebolt/jdbc/connection/FireboltConnection.java

```diff
@@ -58,146 +51,111 @@
-		this.connect();
 	}
 
 	// This code duplication between constructors is done because of back reference: dependent services require reference to current instance of FireboltConnection that prevents using constructor chaining or factory method.
 	@ExcludeFromJacocoGeneratedReport
-	public FireboltConnection(@NonNull String url, Properties connectionSettings) throws SQLException {
+	protected FireboltConnection(@NonNull String url, Properties connectionSettings) throws SQLException {
 		this.loginProperties = extractFireboltProperties(url, connectionSettings);
 		OkHttpClient httpClient = getHttpClient(loginProperties);
 		ObjectMapper objectMapper = FireboltObjectMapper.getInstance();
 
-		this.fireboltAuthenticationService = new FireboltAuthenticationService(new FireboltAuthenticationClient(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients()));
-		this.fireboltGatewayUrlService = new FireboltGatewayUrlService(createFireboltAccountRetriever(httpClient, objectMapper, "engineUrl", GatewayUrlResponse.class));
-		this.httpConnectionUrl = getHttpConnectionUrl(loginProperties);
+		this.fireboltAuthenticationService = new FireboltAuthenticationService(createFireboltAuthenticationClient(httpClient, objectMapper));
+		this.httpConnectionUrl = loginProperties.getHttpConnectionUrl();
 		this.fireboltStatementService = new FireboltStatementService(new StatementClientImpl(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients()));
 
 		this.statements = new ArrayList<>();
 		this.connectionTimeout = loginProperties.getConnectionTimeoutMillis();
 		this.networkTimeout = loginProperties.getSocketTimeoutMillis();
-		this.systemEngine = loginProperties.isSystemEngine();
-		this.fireboltEngineService = new FireboltEngineService(this);
-		this.fireboltAccountIdService = new FireboltAccountIdService(createFireboltAccountRetriever(httpClient, objectMapper, "resolve", FireboltAccount.class));
+	}
+
+	protected abstract FireboltAuthenticationClient createFireboltAuthenticationClient(OkHttpClient httpClient, ObjectMapper objectMapper);
+
+	public static FireboltConnection create(@NonNull String url, Properties connectionSettings) throws SQLException {
+		return createConnectionInstance(url, connectionSettings);
+	}
 
-		this.connect();
+	private static FireboltConnection createConnectionInstance(@NonNull String url, Properties connectionSettings) throws SQLException {
+		switch(getUrlVersion(url)) {
+			case 1: return new FireboltConnectionUserPasswordAuthentication(url, connectionSettings);
+			case 2: return new FireboltConnectionServiceSecretAuthentication(url, connectionSettings);
+			default: throw new IllegalArgumentException(format("Cannot distinguish version from url %s", url));
+		}
+	}
+
+	private static int getUrlVersion(String url) {
+		Pattern urlWithHost = Pattern.compile("jdbc:firebolt://api\\.\\w+\\.firebolt\\.io");
```

Comment:
OldServiceAccountAuthenticationRequest is removed but what is incomplete here? We cannot use user/password or id/secret fields in order to distinguish the connection string version because these information can (and should) be provided as additional properties. The only thing that definitely guarantees that this is v1 is existence of the host in the  JDBC URL.

Reply:
You can have a host in your URL, and still need to connect to V2.
This will happen in case you provide client_id/client_secret as credentials to an old URL format

Reply:
fixed. pls review code and test `FireboltDriverTest.validateConnectionWhenUrlIsValid()`